src/rule_definition_tools.jl

```diff
@@ -249,7 +261,7 @@ function propagation_expr(Δs, ∂s, _conj = false)
         :($(_∂s[1]) * $(Δs[1]))
     end
 
-    return summed_∂_mul_Δs
+    return :($proj($summed_∂_mul_Δs))
 end
```

Comment:
This is a terrifying macro, and the terrible bits only seem to be called about twice... but I think I see where to insert projectors. I think we should probably do this. But it may not belong in this PR, it's one commit. I'd be especially thrilled if someone else would invent a test for this.

Reply:
@sethaxen  can you invent a test using complex numbers for this?

Reply:
The obvious test is something like `gradient(abs2∘*, 1+im, 2)`, with a custom `mul` for which you define a rule. What I meant was that trying to test all corners of this macro seems much messier...

Reply:
we should already have all corners tested.
You just need to worry about the bits that change.

Reply:
I have added some tests. But there is 500 lines of macro code... and if I delete for instance the `.` in `:((muladd).($∂s_i, $Δs_i, $ex))` then nothing fails, I have no idea why it's there. (There seem to be many layers of escaping gensyming which look like they may almost cancel, and macros within macros.)

Not your fault I don't think, but for 1.0 it are you at all tempted to rip it out? If you need `@setup` or `sincos` just write the rule, this would not take 500 lines... and I can count on one hand the use cases hits in ChainRules.

Reply:
Temped yes. but not temped enough to actually do it.
Particularly as it's restrictions garentee that it can be done without augmenting the primal.
Which is what NiLang and Symbolics,jl need.
So if we want to support them (partially) we need a restricted form of rrule and frule that has that feature, and that only operates on scalars.